### PR TITLE
feat: osmosis swap

### PR DESCRIPTION
**Pull request by WaDadidou**
Take attention to the main hook made for this feature : https://github.com/TERITORI/teritori-dapp/blob/feat/osmosis-swap/packages/hooks/useSwap.ts

And to the components made for this feature : https://github.com/TERITORI/teritori-dapp/tree/feat/osmosis-swap/packages/screens/Swap

We will add more currencies in the future (See https://github.com/TERITORI/teritori-dapp/tree/feat/osmosis-swap/packages/networks/osmosis/currencies.ts)

Maybe we need to add some networks modifications (See https://github.com/TERITORI/teritori-dapp/tree/feat/osmosis-swap/packages/networks

### CODE REVIEW

File: packages/components/CurrencyIcon.tsx [hunk 1]
```diff
@@ -8,13 +8,14 @@ export const CurrencyIcon: React.FC<{
   networkId: string;
   denom: string;
   size: number;
-}> = ({ networkId, denom, size }) => {
+  icon?: string;
```
Comment (n0izn0iz) [inline packages/components/CurrencyIcon.tsx:11]:
```suggestion
  fallback?: string;
```

File: packages/components/inputs/TextInputCustom.tsx [hunk 2]
```diff
@@ -82,6 +98,13 @@ export const TextInputCustom = <T extends FieldValues>({
   });
   const inputRef = useRef<TextInput>(null);
 
+  // Passing ref to parent
+  useEffect(() => {
+    if (inputRef.current && setRef) {
+      setRef(inputRef);
+    }
+  }, [setRef]);
```
Comment (n0izn0iz) [inline packages/components/inputs/TextInputCustom.tsx:106]: check https://reactjs.org/docs/forwarding-refs.html

**Review by n0izn0iz — COMMENTED** [review]
partial review

File: packages/hooks/useSwap.ts [hunk 3]
```diff
@@ -0,0 +1,475 @@
+import { coins, isDeliverTxFailure, StdFee } from "@cosmjs/stargate";
+import {
+  calculateAmountWithSlippage,
+  LcdPool,
+  makeLcdPoolPretty,
+  OsmosisApiClient,
```
Comment (omniwired) [inline packages/hooks/useSwap.ts:6]: try importing it like this
```
import {
  calculateAmountWithSlippage,
  LcdPool,
  makeLcdPoolPretty,
  OsmosisApiClient,
  assets as osmosisAssets,
} from "@cosmology/core";
import { useQuery } from "@tanstack/react-query";
```

File: packages/components/inputs/TextInputCustom.tsx (hunk 2 above)
Comment (WaDadidou) [reply packages/components/inputs/TextInputCustom.tsx]: I'm not sure about the types here : https://github.com/TERITORI/teritori-dapp/blob/dd0f8e38bed58c9b85b10dca862ba29383173240/packages/components/inputs/TextInputCustom.tsx#L61
So i used some ugly `any`.
The ref to forward is defined here : https://github.com/TERITORI/teritori-dapp/blob/dd0f8e38bed58c9b85b10dca862ba29383173240/packages/screens/Swap/components/SwapModalSettings.tsx#L71
Wdyt ?

**Review by 3mp8r3 — APPROVED** [review]

**Review by 3mp8r3 — APPROVED** [review]

**Review by omniwired — DISMISSED** [review]

**Review by sakul-budhathoki — APPROVED** [review]

**Pull request merged**